client/starwhale/api/_impl/track/tracker.py

```diff
@@ -0,0 +1,549 @@
+            Tracker._instance = _tracker
+            return _tracker
+
+    @classmethod
+    def end(cls) -> None:
+        """Ends a track.
+        Singleton Tracker instance will be released. If users don't call end function,
+        Tracker will call end function when the process exiting automatically.
+        """
+        with Tracker._lock:
+            if Tracker._instance is None:
+                return
+
+            _inst = Tracker._instance
+            Tracker._instance = None
+
+            _inst._render_manifest()
+            _inst._close_threads()
+
+            if _inst._log_hook is not None:
+                _inst._log_hook.uninstall()
+                _inst._log_hook = None
+
+            atexit.unregister(Tracker.end)
+
+    @classmethod
+    def metrics(
+        cls,
+        *args: t.Dict,
+        step: t.Optional[int] = None,
+        commit: bool = True,
+        **kwargs: float,
+    ) -> Tracker:
+        """Track metrics to Starwhale.
+        Metric value type must be float or convertible-float type. Each call to `metrics` function will trigger
+        internal step counter to be added by 1.
+
+        Arguments:
+            *args: [dict, optional] use dicts to store metrics.
+            **kwargs: [float, optional] use kv to store metrics.
+            step: [int, optional] custom step value.
+            commit: [bool, optional] when `commit` is `false`, the metrics will be aggregated with the same step value
+                until the next `track.metrics(commit=True)` function call. Defaults to `true`.
+
+        Examples:
+        ```python
+        track.metrics(loss=0.99)
+        track.metrics(loss=0.99, accuracy=0.98)
+        track.metrics({"loss": 0.99, "accuracy": 0.98})
+
+        # loss and accuracy use the same step.
+        track.metrics(loss=0.99, commit=False)
+        track.metrics(accuracy=0.98)
+
+        # use custom step value
+        track.metrics(loss=0.99, step=1)
+
+        Returns:
+            A `Tracker` object.
+        ```
+        """
+        if cls._instance is None:
+            raise RuntimeError(_INSTANCE_NOT_INIT_ERROR)
+
+        data = cls._merge_dicts(args, kwargs)
+        cls._instance._log_metrics(data=data, step=step, commit=commit)
+        return cls._instance
+
+    def _log_metrics(
+        self,
+        data: t.Dict[str, float],
+        step: t.Optional[int] = None,
+        commit: bool = False,
+        source: _TrackSource = _TrackSource.USER,
+    ) -> None:
+        if not data:
+            return
+
+        _data = flatten(data, extract_sequence=True)
+        for k, v in _data.items():
+            _data[k] = float(v)
+
+        _step: int = self._metric_step
+        if step is not None:
+            if not isinstance(step, int) or step < 0:
+                raise TypeError(
+                    f"step({step}) is unexpected, the normal step is non-negative integer"
+                )
+            _step = step
+
+        with self._metric_lock:
```

Comment:
Is it not necessary to use this lock if we use _data directly?

Reply:
we use `self._metric_lock` lock to protect `self._metric_step` and `self._uncommitted_metrics`.